src/catalog/database_catalog.cpp

```diff
@@ -30,856 +29,89 @@
-  set_class_pointer_pri_ = classes_->InitializerForProjectedRow(set_class_pointer_oids);
-
-  const std::vector<col_oid_t> set_class_schema_oids{postgres::REL_SCHEMA_COL_OID};
-  set_class_schema_pri_ = classes_->InitializerForProjectedRow(set_class_schema_oids);
-
-  const std::vector<col_oid_t> get_class_pointer_kind_oids{postgres::REL_PTR_COL_OID, postgres::RELKIND_COL_OID};
-  get_class_pointer_kind_pri_ = classes_->InitializerForProjectedRow(get_class_pointer_kind_oids);
-
-  const std::vector<col_oid_t> get_class_schema_pointer_kind_oids{postgres::REL_SCHEMA_COL_OID,
-                                                                  postgres::RELKIND_COL_OID};
-  get_class_schema_pointer_kind_pri_ = classes_->InitializerForProjectedRow(get_class_schema_pointer_kind_oids);
-
-  const std::vector<col_oid_t> get_class_object_and_schema_oids{postgres::REL_PTR_COL_OID,
-                                                                postgres::REL_SCHEMA_COL_OID};
-  get_class_object_and_schema_pri_ = classes_->InitializerForProjectedRow(get_class_object_and_schema_oids);
-  get_class_object_and_schema_prm_ = classes_->ProjectionMapForOids(get_class_object_and_schema_oids);
-
-  // pg_index
-  const std::vector<col_oid_t> pg_index_all_oids{postgres::PG_INDEX_ALL_COL_OIDS.cbegin(),
-                                                 postgres::PG_INDEX_ALL_COL_OIDS.cend()};
-  pg_index_all_cols_pri_ = indexes_->InitializerForProjectedRow(pg_index_all_oids);
-  pg_index_all_cols_prm_ = indexes_->ProjectionMapForOids(pg_index_all_oids);
-
-  const std::vector<col_oid_t> get_indexes_oids{postgres::INDOID_COL_OID};
-  get_indexes_pri_ = indexes_->InitializerForProjectedRow(get_class_oid_kind_oids);
-
-  const std::vector<col_oid_t> delete_index_oids{postgres::INDOID_COL_OID, postgres::INDRELID_COL_OID};
-  delete_index_pri_ = indexes_->InitializerForProjectedRow(delete_index_oids);
-  delete_index_prm_ = indexes_->ProjectionMapForOids(delete_index_oids);
-
-  // pg_type
-  const std::vector<col_oid_t> pg_type_all_oids{postgres::PG_TYPE_ALL_COL_OIDS.cbegin(),
-                                                postgres::PG_TYPE_ALL_COL_OIDS.cend()};
-  pg_type_all_cols_pri_ = types_->InitializerForProjectedRow(pg_type_all_oids);
-  pg_type_all_cols_prm_ = types_->ProjectionMapForOids(pg_type_all_oids);
+void DatabaseCatalog::Bootstrap(const common::ManagedPointer<transaction::TransactionContext> txn) {
+  BootstrapPRIs();
 
-  // pg_language
-  const std::vector<col_oid_t> pg_language_all_oids{postgres::PG_LANGUAGE_ALL_COL_OIDS.cbegin(),
-                                                    postgres::PG_LANGUAGE_ALL_COL_OIDS.cend()};
-  pg_language_all_cols_pri_ = languages_->InitializerForProjectedRow(pg_language_all_oids);
-  pg_language_all_cols_prm_ = languages_->ProjectionMapForOids(pg_language_all_oids);
+  bool UNUSED_ATTRIBUTE retval;
+  retval = TryLock(txn);
+  NOISEPAGE_ASSERT(retval, "Bootstrap operations should not fail");
```

Comment:
If this assert were to ever fail, I suspect we'll need as much help for our future selves as possible...
```suggestion
  NOISEPAGE_ASSERT(retval, "Bootstrap should not fail to get write-lock: another thread grabbed it early? Check recovery logic (most probable cause)");
```

Reply:
thanks